tsdb/batcher.go

```diff
@@ -82,21 +106,46 @@ func (b *PointBatcher) Start() {
 	}()
 }
 
+func (b *PointBatcher) Stop() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	// If not running, nothing to stop.
+	if b.wg == nil {
+		return
+	}
+
+	close(b.stop)
+	b.wg.Wait()
+}
+
 // In returns the channel to which points should be written.
-func (b *PointBatcher) In() chan<- Point { return b.in }
+func (b *PointBatcher) In() chan<- Point {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	return b.in
+}
 
 // Out returns the channel from which batches should be read.
-func (b *PointBatcher) Out() <-chan []Point { return b.out }
+func (b *PointBatcher) Out() <-chan []Point {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	return b.out
+}
 
 // Flush instructs the batcher to emit any pending points in a batch, regardless of batch size.
 // If there are no pending points, no batch is emitted.
 func (b *PointBatcher) Flush() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
 	b.flush <- struct{}{}
 }
 
 // Stats returns a PointBatcherStats object for the PointBatcher. While the each statistic should be
 // closely correlated with each other statistic, it is not guaranteed.
 func (b *PointBatcher) Stats() *PointBatcherStats {
+	b.mu.RLock()
```

Comment:
I don't believe this locking is not necessary. Each entry of the new `stats` object is loaded using an atomic load instruction.